### DISCUSSION TITLE
Where should I put additional python src files that pipelines depend on?

### DISCUSSION THREAD

**Discussion by alexeyabel** [kedro-org/kedro#1665] [Idea]
I have a custom package that implements a model calculation that my pipelines use (depend on). Where should I put it in the kedro folder structure?

**variant 1**
```
kedro_project/
└── src
    └── kedro_project
        ├── my_custom_package
        │   └── __init__.py
        └── pipelines
```
**or variant 2**
```
kedro_project/
└── src
    ├── kedro_project
    │   └── pipelines
    └── my_custom_package
        └── __init__.py
```
I was not able to find an example for this. Or is this an anti-pattern and my custom package should be completely separated from the kedro hirarchy?

**Comment by datajoely** [comment]
Variant 2 is the preferred approach, your business logic should not be tightly coupled with your pipeline logic :)

**Comment by deepyaman** [comment]
> Or is this an anti-pattern and my custom package should be completely separated from the kedro hirarchy?

I wouldn't go so far as to say it's an anti-pattern, but it does warrant additional thought. By default, the Kedro project template is intended for developing a single package. The `src/setup.py` file contains instructions for packaging it.

If you introduce a second package, especially if you want to publish it, it will need it's own packaging instructions. It's also unclear now, without looking at the contents of `src/setup.py`, which package does it provide instructions for? If you don't do this, you're simply relying on local imports.

You should either have a separate repository structure, or you can restructure your repository as a monorepo. There are a number of documented (Kedro-agnostic) pros/cons of each approach.
